pandas/core/ops/array_ops.py

```diff
@@ -132,3 +153,232 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
         result = masked_arith_op(left, right, op)
 
     return missing.dispatch_fill_zeros(op, left, right, result)
+
+
+def arithmetic_op(
+    left: Union[np.ndarray, ABCExtensionArray],
+    right: Any,
+    op,
+    str_rep: str,
+    eval_kwargs: Dict[str, str],
+):
+    """
+    Evaluate an arithmetic operation `+`, `-`, `*`, `/`, `//`, `%`, `**`, ...
+
+    Parameters
+    ----------
+    left : np.ndarray or ExtensionArray
+    right : object
+        Cannot be a DataFrame or Index.  Series is *not* excluded.
+    op : {operator.add, operator.sub, ...}
+        Or one of the reversed variants from roperator.
+
+    Returns
+    -------
+    ndarrray or ExtensionArray
+        Or a 2-tuple of these in the case of divmod or rdivmod.
+    """
+
+    from pandas.core.ops import (
+        maybe_upcast_for_op,
+        should_extension_dispatch,
+        dispatch_to_extension_op,
+    )
+
+    keep_null_freq = isinstance(
+        right,
+        (
+            ABCDatetimeIndex,
+            ABCDatetimeArray,
+            ABCTimedeltaIndex,
+            ABCTimedeltaArray,
+            Timestamp,
+        ),
+    )
+
+    # NB: We assume that extract_array has already been called on `left`, but
+    #  cannot make the same assumption about `right`.  This is because we need
+    #  to define `keep_null_freq` before calling extract_array on it.
```

Comment:
Would another option be to define keep_null_freq beforehand and pass that to arithmetic_op? 
Then both extract_arrays can be done beforehand?

Reply:
I considered that, but that makes things more complicated once we start calling these from multiple places (mainly Block-wise ops, to a lesser extent PandasArray).